File: 5kyu/factorial_decomposition.cpp
#include <string>
#include <iostream>

std::string decomp(int n) {
    bool primes[n + 1];
    std::fill_n(primes, n + 1, true);
    for (int i = 2; i * i <= n; i++) {
        if (primes[i] == true)
            for (int j = i * 2; j <= n; j += i)
                primes[j] = false;
    }
    std::string s;
    for (int i = 2; i <=n; i++) {
        if (primes[i]) {
            int exp = 0, tmp = n;
            while (tmp) {
                tmp /= i;
                exp += tmp;
            }
            if (exp) {
                if (i != 2)
                    s += " * ";
                s += std::to_string(i);
                if (exp > 1)
                    s += "^" + std::to_string(exp);
            }
        }
    }
    return s;
}

File: 5kyu/factorial_decomposition.py
def decomp(n):
    primes = sieve(n)
    count = {i: 0 for i in primes}
    for prime in sieve(n):
        exp, tmp = 0, n
        while tmp:
            tmp //= prime
            exp += tmp
        count[prime] = exp
    return ''.join(f'{str(k) + ("^" + str(v) if v > 1 else "")} * '
                   for k, v in count.items() if v)[:-3]


def sieve(n):
    primes = [True] * (n + 1)
    primes[0] = primes[1] = False
    i = 2
    while i * i <= n:
        if primes[i]:
            for non_primes in range(i * i, n + 1, i):
                primes[non_primes] = False
        i += 1
    return (i for i, b in enumerate(primes) if b)
